src/java/org/apache/cassandra/locator/Replicas.java

```diff
@@ -50,6 +50,30 @@
     public abstract int size();
     protected abstract Collection<Replica> getUnmodifiableCollection();
 
+
+    public boolean equals(Object o)
```

Comment:
Sorry I should have been clearer. We would need AbstractReplicaList and AbstractReplicaSet to each implement their own equals and hash code. If you look at Java's AbstractList and AbstractSet you will see they implement their own equals and hashCode where the two aren't ever equal.

And I think it's not just cargo culting to also do it that way. I think it makes sense they should never be equal because sets have unpredictable iteration order and it leads to bugs to have it work only some of the time. Sets  also use a hash code specific to sets which is the sum of hash codes of the elements which makes it order insensitive.

Reply:
Good point on set/list equality. Reverted.

>  We should never return a Replicas that isn't either an instance of either ReplicaSet or ReplicaList 

Why not? The methods that return the immutable containers are basically just returning transformed views of their arguments. The fact that they return generic `ReplicaCollection` instances is a pretty clear indicator that you can't expect them to behave like lists or sets.

Reply:
> If you look at how java.util.Collections and Guava do it you can't actually construct something that isn't either a set or a list.

The objects returned by com.google.common.collect.Iterables (which these methods are imitating) don't define equals or hashCode. 

The issue here might be that I'm thinking of these objects as iterables, and you seem to be thinking of them as  collections. Maybe this is more a naming issue?

Reply:
Sure if they don't implement they implement Iterable instead of Collection I think it's fine for them to not be an implementation of a collection (list, set, multi-set).

Interesting side note, java.util.AbstractQueue doesn't implement equals! Neither does say ArrayBlockingQueue or ArrayDeque.